ObjCThemis: Secure Cell API update

ilammy opened the pull request:
Update Secure Cell API for Objective-C and Swift as a preparation to passphrase support. This has been described in RFC 3.5.

This is a non-functional change, just making the API more consistent and easy to use, with a straightforward way to add passphrase API soon.

## User notes

### Swift

All modes now use consistent naming of encryption/decryption methods:

```swift
let encrypted = try cell.encrypt(message, context: nil)

let decrypted = try cell.decrypt(encrypted)
```

They accept `Data` as their first unnamed argument, followed an optional `context:` argument (except for Context Imprint mode which requires associated context). Processed data is returned as `Data`. Errors are reported with Swift exceptions.

#### Token Protect mode improvements

Token Protect mode returns a helper object with encrypted message and authentication token:

```swift
let result = try cell.encrypt(message)

print("Encrypted message:    ", result.encrypted)
print("Authentication token: ", result.token)
```

Encrypted message and token are accepted separately for decryption, no longer requiring to fill `TSCellTokenEncryptedData` manually:

```swift
let decrypted = try cell.decrypt(encrypted, token: token)
```

New API is bridged with Swift's `Data` directly, no longer requiring manual conversion from/to `NSMutableData`.

### Objective-C

All modes now use consistent naming of encryption/decryption methods:

```objective-c
NSData *encrypted = [cell encrypt:message];

NSError *error;
NSData *decrypted = [cell decrypt:encrypted
                          context:nil
                            error:&error];
```

They accept `NSData` as their first argument, followed an optional `context:` argument (except for Context Imprint mode which requires associated context), followed by an optional `error:` out-parameter. Processed data is returned as `NSData`. Errors are reported as `NSError` objects, if requested.

There are convenience overloads with omitted `context` parameters if you don't need them, and omitted `error:` parameters if you're not interested in error values (`nil` is returned in case of error).

#### Token Protect mode improvements

Token Protect mode returns a helper result object with encrypted message and authentication token:

```objective-c
TSCellTokenEncryptedResult *result = [cell encrypt:message];

NSLog(@"Encrypted message:    %@", result.encrypted);
NSLog(@"Authentication token: %@", result.token);
```

Encrypted message and token are accepted separately for decryption, no longer requiring to fill `TSCellTokenEncryptedData` manually:

```objective-c
NSData *decrypted = [cell decrypt:encrypted token:token];
```

Note that new helper object `TSCellTokenEncryptedResult` represents only encryption result. It is immutable and uses `NSData` for fields.

### Deprecated API

Old encryption/decryption API is declared deprecated and its use is discouraged. There are no plans to remove this API in the future, it will be retained for compatibility (probably until 1.0).

<details>
<summary>Complete list of deprecated API</summary>

Swift:

<table>
  <tr><th>Mode</th><th>Deprecation</th><th>Replacement</th></tr>
  <tr>
    <td rowspan=2><code>TSCellSeal</code></td>
    <td><code>wrap(_:, context:)</code><br/><code>wrap</code></td>
    <td><code>encrypt(_:, context:)</code><br/><code>encrypt</code></td>
  </tr>
  <tr>
    <td><code>unwrapData(_:, context:)</code><br/><code>unwrapData</code></td>
    <td><code>decrypt(_:, context:)</code><br/><code>decrypt</code></td>
  </tr>
  <tr>
    <td rowspan=2><code>TSCellToken</code></td>
    <td><code>wrap(_:, context:)</code><br/><code>wrap</code></td>
    <td><code>encrypt(_:, context:)</code><br/><code>encrypt</code></td>
  </tr>
  <tr>
    <td><code>unwrapData(_:, context:)</code><br/><code>unwrapData</code></td>
    <td><code>decrypt(_:, token:, context:)</code><br/><code>decrypt(_:, token:)</code></td>
  </tr>
  <tr>
    <td rowspan=2><code>TSCellContextImprint</code></td>
    <td><code>wrap(_:, context:)</code><br/><code>wrap</code></td>
    <td><code>encrypt(_:, context:)</code><br/><code>encrypt</code></td>
  </tr>
  <tr>
    <td><code>unwrapData(_:, context:)</code><br/><code>unwrapData</code></td>
    <td><code>decrypt(_:, context:)</code><br/><code>decrypt</code></td>
  </tr>
</table>

Objective-C:

<table>
  <tr><th>Mode</th><th>Deprecation</th><th>Replacement</th></tr>
  <tr>
    <td rowspan=2><code>TSCellSeal</code></td>
    <td><code>wrapData:context:error:</code><br><code>wrapData:error:</code></td>
    <td><code>encrypt:context:error:</code><br><code>encrypt:error:</code></td>
  </tr>
  <tr>
    <td><code>unwrapData:context:error:</code><br><code>unwrapData:error:</code></td>
    <td><code>decrypt:context:error:</code><br><code>decrypt:error:</code></td>
  </tr>
  <tr>
    <td rowspan=2><code>TSCellToken</code></td>
    <td><code>wrapData:context:error:</code><br><code>wrapData:error:</code></td>
    <td><code>encrypt:context:error:</code><br><code>encrypt:error:</code></td>
  </tr>
  <tr>
    <td><code>unwrapData:context:error:</code><br><code>unwrapData:error:</code></td>
    <td><code>decrypt:token:context:error:</code><br><code>decrypt:token:error:</code></td>
  </tr>
  <tr>
    <td rowspan=2><code>TSCellContextImprint</code></td>
    <td><code>wrapData:context:error:</code><br><code>wrapData:error:</code></td>
    <td><code>encrypt:context:error:</code><br><code>encrypt:error:</code></td>
  </tr>
  <tr>
    <td><code>unwrapData:context:error:</code><br><code>unwrapData:error:</code></td>
    <td><code>decrypt:context:error:</code><br><code>decrypt:error:</code></td>
  </tr>
</table>

</details>

## Technical notes

This PR includes significant overhaul of unit test suite for ObjCThemis. It improves coverage and debuggability of ObjCThemis via tests. The test suite also includes compatibility tests.

There are a lot of tedious and repeatative changes there. All modes have roughly the same test suite but with *slight* differences. Objective-C and Swift test suite are identical (Swift one is a translation). However, they are important to pinpoint API details.

There were plans to improve Token Protect encryption API in Swift to return _tuples_:

```swift
let (encrypted, token) = try cell.encrypt(message)

print("Encrypted message:    ", encrypted)
print("Authentication token: ", token)
```

Unfortunately, this turned out to be impossible at the moment due to lack of stable Swift ABI.

## Checklist

- [X] Change is covered by automated tests
- [X] The [coding guidelines] are followed
- [X] Public API has proper documentation
- [X] ~~Example projects and code samples are up-to-date~~ (no, will update separately)
- [X] Changelog is updated

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

P.S. I need to slow down because @vixentael capacity is extremely limited recently.

P.P.S. Why did I even started this?.. It was a perfectly fine API. Yeah, it was _a bit inconvenient_ to use, but it did work. We all know that worse is better, right? Whatever. Here you go.

vixentael commented on src/wrappers/themis/Obj-C/objcthemis/scell_context_imprint.h:
```diff
@@ -47,28 +47,91 @@ NS_ASSUME_NONNULL_BEGIN
+ * Data is encrypted and context is cryptographically mixed with the data, but not included
+ * into the resulting encrypted message. You will have to provide the same context again
+ * during decryption. Usually this is some plaintext data associated with encrypted data,
+ * such as database row number, protocol message ID, etc.
```
```suggestion
 * Data is encrypted and context is cryptographically mixed with the data, but not included
 * into the resulting encrypted message. You will have to provide the same context again
 * during decryption. Usually this is some plaintext data associated with encrypted data,
 * such as database row number, protocol message ID, etc.
 * Length preserving encryption mode without enforced integrity check.
 * @link https://docs.cossacklabs.com/pages/secure-cell-cryptosystem/#context-imprint-mode
```

ilammy replied:
I believe this better be a part of class description rather than encryption function call. It's more likely that people will type `TSCell`, see four possible construction options, and then go off reading their descriptions.

https://github.com/cossacklabs/themis/blob/83f25359a300aa7b9e91b75d8f35b5304c2e659d/src/wrappers/themis/Obj-C/objcthemis/scell_context_imprint.h#L32-L43

I did not update them for some reason, but now that you've mentioned it, it's probably a good time to do it.

vixentael commented on src/wrappers/themis/Obj-C/objcthemis/scell_seal.h:
```diff
@@ -47,43 +47,163 @@ NS_ASSUME_NONNULL_BEGIN
+ * The context, if provided, is cryptographically mixed with the data, but is not included
+ * into the resulting encrypted message. You will have to provide the same context again
+ * during decryption. Usually this is some plaintext data associated with encrypted data,
+ * such as database row number, protocol message ID, etc.
```
```suggestion
 * such as database row number, protocol message ID, etc.
 * The most popular mode, enforces integrity, length of encrypted data is larger than input text length.
 * @link https://docs.cossacklabs.com/pages/secure-cell-cryptosystem/#seal-mode
```

vixentael commented on src/wrappers/themis/Obj-C/objcthemis/scell_token.h:
```diff
@@ -59,44 +78,168 @@ NS_ASSUME_NONNULL_BEGIN
+ * The context, if provided, is cryptographically mixed with the data, but it not included
+ * into the resulting encrypted message. You will have to provide the same context again
+ * during decryption. Usually this is some plaintext data associated with encrypted data,
+ * such as database row number, protocol message ID, etc.
```
```suggestion
 * such as database row number, protocol message ID, etc.
 * @link https://docs.cossacklabs.com/pages/secure-cell-cryptosystem/#token-protect-mode
```

vixentael commented:
> P.P.S. Why did I even started this?.. It was a perfectly fine API.

lololol, i don't know why you started this :D
but i know, that there're other languages that use `wrap/unwrap` syntax.

———————-

As usual, please make sure that we have tasks for:
- updating docs on docserver with new API
- updating example apps with new API

ilammy commented:
> You are API freak :D

The RFCs were only a warning! Now I'll be renaming a method every month until you comply! /s

> but i know, that there're other languages that use wrap/unwrap syntax.

Well, in fact, *that* was the reason and the comment is more like an old-man-shouts-at-a-cloud rant.

When I was reviewing Secure Cell API across all platforms I found that Objective-C, Java, and for some reason Go use this idiosyncratic `unwrap` naming while all other languages *including* the core library use "encrypt" and "decrypt" (the same goes for documentation).

- With Go it's somewhat jarring, considering that everywhere across Acra we call outer methods encrypt and decrypt but deep inside them it's "wrap" and "unwrap" because of reasons.\
- With Java we'll have to deprecate API anyway to discourage passing `String` as master key, and I don't want the users to see 5 deprecated methods and 1 correct one, with all of them called `wrap`.
- With Swift, the name translation turns Objective-C method `wrapData:error:` into `wrap()` but `unwrapData:error:` into `unwrapData()` because IDK why. It's kinda asymmetric that users have to `wrap` their data but then `unwrapData` their data back.
- With Objective-C (and Swift) I planned to improve Token Protect API to not require an intermediate object and reasons similar to Java it would be cleaner to add a new set of API than keep multiple deprecations in autocomplete list.

So you could say these are pretty silly OCD reasons for renaming (and reeducating the users), but hey, we're now more consistent and the old API is not going anywhere. \**cranks up [sunk cost fallacy](https://en.wikipedia.org/wiki/Escalation_of_commitment) up to 11*\*

----

BTW, it seems that Bitrise did *not* test this PRs. I'm kinda wary to merge it without that, so I'll see how to talk it into testing...

ilammy commented:
@vixentael, I've updated class description docs, cleaned up formatting there, and added the links to documentation that you suggested. Please give it a look.

It seems that modern Xcode has dropped support for HeaderDoc completely. Advanced tags like `@link` or `@ref` are not working anymore, unfortunately. Some limited subset is still there, but it seems Apple wants to send a strong message “please migrate to Swift and use Markdown there”.

Also, this push has woken up Bitrise so we'll finally see if the tests pass there with the changes.

vixentael commented:
> please migrate to Swift

and enjoy failing code with every new version of Swift lol

ilammy commented:
> > please migrate to Swift
>
> and enjoy failing code with every new version of Swift lol

Well, if you make a *slight change* – Swift 5 ⟹ Swift 0.5 – then it suddenly makes a lot of sense! Give them some compassion, they've just lost a valuable number and still can't find it.
